core/trino-main/src/main/java/io/trino/sql/planner/TranslationMap.java

```diff
@@ -309,6 +309,19 @@ public Expression rewriteDereferenceExpression(DereferenceExpression node, Void
                                 new LongLiteral(Long.toString(index + 1))));
             }
 
+            @Override
+            public Expression rewriteSubscriptExpression(SubscriptExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                Optional<Expression> mapped = tryGetMapping(node);
+                if (mapped.isPresent()) {
+                    return coerceIfNecessary(node, mapped.get());
+                }
+
+                Expression rewrittenBase = treeRewriter.rewrite(node.getBase(), context);
+                // Do not rewrite subscript index into symbol. Index is required to be a literal.
+                return coerceIfNecessary(node, new SubscriptExpression(rewrittenBase, node.getIndex()));
```

Comment:
Subscript must be literal for `Row` type, but subscript can be used with arrays and maps too.